src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypeStyleDiagnosticAnalyzerBase.cs

```diff
@@ -62,9 +48,7 @@ private void HandleVariableDeclaration(SyntaxNodeAnalysisContext context)
 
             var typeStyle = Helper.AnalyzeTypeName(
                 declaredType, semanticModel, optionSet, cancellationToken);
-            if (!typeStyle.IsStylePreferred ||
-                typeStyle.Severity.WithDefaultSeverity(DiagnosticSeverity.Hidden) >= ReportDiagnostic.Hidden ||
```

Comment:
what is this hidden diagnostic check? is IDE used to have different behavior based on severity?

Reply:
The original intent is avoiding a case where a user tries to apply a Fix All to a warning or error, and the Fix All operation tries to fix a bunch of hidden diagnostics that the user intended to ignore. Multiple mitigations for this scenario are already in place, but if desired we could implement additional protection by filtering the diagnostics selected for Fix All to only include diagnostics at the severity chosen for Fix All or more severe (e.g. Fix All on an Error diagnostic would fix all Error severity; Fix All on a Warning diagnostic would fix all Error and Warning severity; etc.). The easiest approach would be the one seen here, where Fix All simply means to fix all diagnostics.

Reply:
so, are we saying whether fix all for diagnostics that have different severities due to editorconfig between a different portion of directories/files in a solution, each "fix all provider" have its own behavior?

is it something we want to generalize in the engine?

sounds like a design question?

...

should fix all engine fix only diagnostics with same (severity, id, equivalent key) rather than (id, equivalent) or is it something each fix all provider decide?

Reply:
also about Full Solution Analysis. I think I sent telemetry data a few days ago on IDE options. FSA is 3rd most popular options user changes among all Roslyn options.

Reply:
in case people are curious. 
First most popular option is naming preferences (http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Simplification/SimplificationOptions.cs,103)

Second most popular option is placing system first option
(http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Editing/GenerationOptions.cs,9)

Reply:
> The original intent is avoiding a case where a user tries to apply a Fix All to a warning or error, and the Fix All operation tries to fix a bunch of hidden diagnostics that the user intended to ignore.

I think the underlying issue is naming style analyzers have different dynamic diagnostic descriptors for different symbol kind with possibly different severities, and all of them use the same diagnostic ID. Ideally:

1. All diagnostic instances in a compilation with a specific diagnostic ID *must* have same _effective severity_.
2. Performing FixAll on a specific diagnostic, with an explicitly chosen scope (document, project, solution) *must* fix all diagnostic instances with that ID in the requested scope. Effective severity should never play a part in determining fix all behavior.

Reply:
@mavasani with editorconfig, isn't user now possible to set different severity for same diagnostic id in different part of directory? so effective severity can different per directory or file I guess now?

for 2, I think right fix might be making fix all works with error list so that user can select bunch of errors in error list and invoke fix all rather than we having bunch of heuristic that might make sense in one situation and doesn't in other context. and what end result user wants is different on every occasion based on why user is invoking it.

with code clean up framework we have now ("fix all" custom set of diagnostics + multiple fixer), we might be able to extend it a bit more to support error list.


complication on error list over code clean up was that, code clean up only care about diagnostic/fixer kind rather than each specific instance of diagnostic, so it was easy to track between each fixer, but error list is about specific instance of diagnostic, so a bit harder to track on each fixes.

Reply:
> with editorconfig, isn't user now possible to set different severity for same diagnostic id in different part of directory? so effective severity can different per directory or file I guess now?

Yes, I always forget the side effects of editorconfig's directory based configurations. I wonder if we need a change in FixAll experience to allow users to choose between fixing all diagnostics with (a) same effective severity or (b) same or more severe effective severity or (c) regardless of effective severity.